packages/compartment-mapper/src/types.js

```diff
@@ -478,29 +499,71 @@ export {};
+ * @template [BuiltinsPolicyItem=void] Additional types for a builtin policy item
+ * @template [ExtraOptions=unknown] Additional options
  * @typedef {object} PackagePolicy
  * @property {string} [defaultAttenuator] - The default attenuator.
  * @property {PolicyItem<PackagePolicyItem>} [packages] - The policy item for packages.
  * @property {PolicyItem<GlobalsPolicyItem>|AttenuationDefinition} [globals] - The policy item or full attenuation definition for globals.
  * @property {PolicyItem<BuiltinsPolicyItem>|NestedAttenuationDefinition} [builtins] - The policy item or nested attenuation definition for builtins.
  * @property {boolean} [noGlobalFreeze] - Whether to disable global freeze.
+ * @property {ExtraOptions} [options] - Any additional user-defined options can be added to the policy here
  */
 
 /**
  * An object representing a base policy.
- * @template [PackagePolicyItem=void]
- * @template [GlobalsPolicyItem=void]
- * @template [BuiltinsPolicyItem=void]
+ *
+ * @template [PackagePolicyItem=void] Additional types for a package policy item
+ * @template [GlobalsPolicyItem=void] Additional types for a global policy item
+ * @template [BuiltinsPolicyItem=void] Additional types for a builtin policy item
+ * @template [ExtraOptions=unknown] Additional package-level options
  * @typedef {object} Policy
- * @property {Record<string, PackagePolicy<PackagePolicyItem, GlobalsPolicyItem, BuiltinsPolicyItem>>} resources - The package policies for the resources.
+ * @property {Record<string, PackagePolicy<PackagePolicyItem, GlobalsPolicyItem, BuiltinsPolicyItem, ExtraOptions>>} resources - The package policies for the resources.
  * @property {string} [defaultAttenuator] - The default attenuator.
- * @property {PackagePolicy<PackagePolicyItem, GlobalsPolicyItem, BuiltinsPolicyItem>} [entry] - The package policy for the entry.
+ * @property {PackagePolicy<PackagePolicyItem, GlobalsPolicyItem, BuiltinsPolicyItem, ExtraOptions>} [entry] - The package policy for the entry.
  */
 
 /**
  * Any object. All objects. Not `null`, though.
  * @typedef {Record<PropertyKey, any>} SomeObject
  */
+
+/**
+ * Any {@link PackagePolicy}
+ *
+ * @typedef {PackagePolicy<any, any, any, any>} SomePackagePolicy
+ */
+
+/**
+ * Any {@link Policy}
+ *
+ * @typedef {Policy<any, any, any, any>} SomePolicy
+ */
+
+/**
+ * A primitive
+ *
+ * Lifted from {@link https://npm.im/type-fest type-fest}
+ *
+ * @typedef {null|undefined|string|number|boolean|symbol|bigint} Primitive
+ * @see {@link LiteralUnion}
+ */
+
+/**
+ * Allows creation of a union of a primitive and a literal type which is not loosened to the primitive.
+ *
+ * Aids type hints in IDEs.
+ *
+ * Lifted from {@link https://npm.im/type-fest type-fest}
+ *
+ * @template LiteralType The literal type
+ * @template {Primitive} PrimitiveType The primitive type
+ * @typedef {LiteralType | (PrimitiveType & Record<never, never>)} LiteralUnion
```

Comment:
Stunning.

Reply:
Please lift the whole thing, since it explains why it's necessary and when it will not longer be necessary. 

Consider separating a `type-fest.d.ts` or `utils.d.ts` that has these utility types that aren't of compartment-mapper per se.

> I welcome a comment from @turadg whether this should become integral to our style for enum literals.

Union of literals is how in agoric-sdk we specify possible values for enums. But that doesn't require this new type. This one is only for "when a union type of a primitive type is combined with literal types".

Reply:
@turadg Lifting the whole thing from `type-fest` would involve several files.  What about: move this to `type-fest.d.ts`, and copy a bunch of the relevant comments over?

Reply:
@turadg Also, would you still want this file in `compartment-mapper`?

Reply:
> Lifting the whole thing from type-fest would involve several files

@boneskull , I just meant the whole jsdoc (basically https://github.com/sindresorhus/type-fest/blob/main/source/literal-union.d.ts)

> move this to type-fest.d.ts, and copy a bunch of the relevant comments over?

Works for me. Slight preference for `utils.d.ts` to allow for more like this. You might need an empty `.js` twin; our linter config has gotten tripped up before I think.

> would you still want this file in compartment-mapper?

It's so small I think it's fine to vendor in each package that uses it. TypeScript is structural so there won't be any difference in its type resolution.

Reply:
@turadg AFAICT the TS configuration won't consider the validity of a `.d.ts` file when type-checking.  I'm not sure of a way to force the checking of a single `.d.ts` file.  However, a `.ts` or `.js` file _would_ be type-checked.  To that end, it may be more straightforward to leave these definitions as `@typedef`s in a `.js` file.

Reply:
You're right that it won't typecheck the `.d.ts` itself (because of skipLibCheck) but it would observe the type when imported into another module. We do this all the time in agoric-sdk. https://github.com/Agoric/agoric-sdk/blob/master/docs/typescript.md#dts-modules

But I don't want to slow you down. Factor it however you like. My only "please" was to include the full JSDoc comment. The other points were to "consider".